From 9db60bf91974144ef2322c8976b657e2e71bd7b7 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:33 -0400
Subject: [fs] btrfs: new function read_extent_buffer_to_user

Message-id: <1407179109-13714-387-git-send-email-sandeen@redhat.com>
Patchwork-id: 87263
O-Subject: [RHEL7.1 PATCH 386/422] btrfs: new function read_extent_buffer_to_user
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Gerhard Heift <gerhard@heift.name>

commit 550ac1d85ef99f3390a6ea87c70b7683647f6110
Author: Gerhard Heift <gerhard@heift.name>
Date:   Thu Jan 30 16:24:01 2014 +0100

    btrfs: new function read_extent_buffer_to_user

    This new function reads the content of an extent directly to user memory.

    Signed-off-by: Gerhard Heift <Gerhard@Heift.Name>
    Signed-off-by: Chris Mason <clm@fb.com>
    Acked-by: David Sterba <dsterba@suse.cz>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/extent_io.c b/fs/btrfs/extent_io.c
index 476a71b..1da3e6f 100644
--- a/fs/btrfs/extent_io.c
+++ b/fs/btrfs/extent_io.c
@@ -5077,6 +5077,43 @@ void read_extent_buffer(struct extent_buffer *eb, void *dstv,
  }
 }
 
+int read_extent_buffer_to_user(struct extent_buffer *eb, void __user *dstv,
+   unsigned long start,
+   unsigned long len)
+{
+ size_t cur;
+ size_t offset;
+ struct page *page;
+ char *kaddr;
+ char __user *dst = (char __user *)dstv;
+ size_t start_offset = eb->start & ((u64)PAGE_CACHE_SIZE - 1);
+ unsigned long i = (start_offset + start) >> PAGE_CACHE_SHIFT;
+ int ret = 0;
+
+ WARN_ON(start > eb->len);
+ WARN_ON(start + len > eb->start + eb->len);
+
+ offset = (start_offset + start) & (PAGE_CACHE_SIZE - 1);
+
+ while (len > 0) {
+  page = extent_buffer_page(eb, i);
+
+  cur = min(len, (PAGE_CACHE_SIZE - offset));
+  kaddr = page_address(page);
+  if (copy_to_user(dst, kaddr + offset, cur)) {
+   ret = -EFAULT;
+   break;
+  }
+
+  dst += cur;
+  len -= cur;
+  offset = 0;
+  i++;
+ }
+
+ return ret;
+}
+
 int map_private_extent_buffer(struct extent_buffer *eb, unsigned long start,
           unsigned long min_len, char **map,
           unsigned long *map_start,
diff --git a/fs/btrfs/extent_io.h b/fs/btrfs/extent_io.h
index 8b63f2d..15ce5f2 100644
--- a/fs/btrfs/extent_io.h
+++ b/fs/btrfs/extent_io.h
@@ -304,6 +304,9 @@ int memcmp_extent_buffer(struct extent_buffer *eb, const void *ptrv,
 void read_extent_buffer(struct extent_buffer *eb, void *dst,
    unsigned long start,
    unsigned long len);
+int read_extent_buffer_to_user(struct extent_buffer *eb, void __user *dst,
+          unsigned long start,
+          unsigned long len);
 void write_extent_buffer(struct extent_buffer *eb, const void *src,
     unsigned long start, unsigned long len);
 void copy_extent_buffer(struct extent_buffer *dst, struct extent_buffer *src,
-- 
1.7.1